### DISCUSSION TITLE
dataset question

### DISCUSSION THREAD

**Discussion by rtwod2** [pyg-team/pytorch_geometric#4405] [Q&A]
Hi all, 

I'm having a dataset of kernels, for which kernel there are more than one corresponding graphs. This depends on input bytes for example. I'm doing a graph regression using PyTorch geometric. For training I have 150 graphs, for validation 120 and for testing 39 graphs. Is this number of graphs enough?

I'm not sure if this is a problem, but for example in test set, the number of 39 is out of 3 kernels with different input bytes. 

Thank u very much for your time and consideration.

**Comment by rusty1s** [comment]
This is totally hard to say. A similar sized graph dataset is the `MUTAG` dataset of the `TUDataset` suite (188 graphs in total), which sees heavy fluctuations in a 10-fold cross validation scenario, but works totally fine on average. I think it's best to just try it out and see how it performs in practice :)

**Reply by rtwod2** [reply]
Yes, you're right. As I mentioned previously this is a regression problem, and I'm predicting 2 targets. I have tested my model and I have the following problem.
After 300 epochs, training loss is 0.41, validation loss is 0.49 and then just for testing I have test loss 0.22. The thing is that even though for training it seems okay (after visualization and checking targets labels numbers) for validation and especially for testing the values predicted tend to be the same, and maybe for some target/label pair works, but this is not the general case. Also, in loss plot I see some overfitting.
![train_valid_loss_new](https://user-images.githubusercontent.com/98180692/161533408-30a22580-08d6-4adb-a1db-d15eed971aaf.png)

What I found is that when for example I'm calculating something like an accuracy metric (just to check the predictions using metrics), I found that having y.shape [1,2] like values [-0.0582, -0.6618] (normalized labels) there is a prediction [-0.0573, -0.1186]. This is close only for the first target y[:,0] and not for the y[:,1], so in this batch row I'll have [True, False], and the accuracy takes this as a correct one, which is something that I don't want to (I would prefer [True, True]). 
So to make a point out of the above, I'm thinking if the L1Loss function that I'm using has the same problem, not taking account both of them. If yes, do you know how to change the L1Loss to take into account both targets? 
Would it be the reason that I'm having same values in the validation and mostly in the test set?

Thank u once again. You're so helpful.

**Reply by rusty1s** [reply]
`L1Loss` should already take into account both. You can also flatten the vector before computing it. Not sure how you compute accuracy from a regression problem though. MAE or MSE sound like better options to me :)

**Reply by rtwod2** [reply]
I'm not sure if flattening is the way to go, because if you look the above example with the numbers 
> y.shape [1,2] like values [-0.0582, -0.6618] (normalized labels) there is a prediction [-0.0573, -0.1186] 

I want to know that prediction -0.0573 is a pair with the prediction -0.1186.

I don't compute an actual accuracy, because as a regression problem there's no reason to do it. But what am I doing is at first calculating the distance between abs(predictions-targets) and if this distance is lower than a small percentage (10%) then I'm accepting the values. I just wanted to visualize with numbers my results. 

After visualizing my results, I found that the y[:,1] has no good predictions.
What I have tried though is to use L1Loss with reduction=none, so by this I have the loss for each target/prediction in my batch separately, like below

**loss_eval_all** is torch.Size([16, 2]) 
tensor([[0.0358, 0.4436],
        [0.0304, 0.2045],
        [0.0305, 0.0331],
        [0.0292, 0.2699],
        [0.0288, 0.5115],
        [0.0302, 0.7486],
        [0.0286, 0.9883],
        [0.0272, 1.2269],
        [0.0273, 1.4642],
        [0.0313, 1.7011],
        [0.0293, 1.9406],
        [0.0283, 2.1810],
        [0.0320, 2.4158],
        [0.0299, 2.6577],
        [0.0300, 2.8950],
        [0.0315, 3.1317]])

Then I'm computing the mean loss for each column and I have for example this thing:
     **loss_eval** torch.Size([2])     
      tensor([0.0300, 1.4259])
During each epoch, I'm gathering all these mean corresponding losses from each step in my epoch, and I take the corresponding average loss in the end of the epoch.
My goal of doing this, is to train the model equally for the y[:,1] target as well. So train against two losses. I also provide the gradients in the right shape when calling backward during training, like this (loss_train.backward(torch.ones_like(loss_train))).

The problem is that when I'm trying to run this, after 25 epochs the training is killed and I can't find out the reason why. I noticed that the MEM usage is of 82% and the CPU usage is of 574% which is enormous. Would it be that the 2-value loss tensor passing when calling the .backward() is the cause of making my system unresponsive?
 
Do you think that having 2values as loss per target is a good approach? 

For once again I'm grateful for your help.

**Reply by rtwod2** [reply]
In addition to the above questions I add one more.

![normalizedexecutiondistribution_onlygpu](https://user-images.githubusercontent.com/98180692/161988879-fef6f1d7-c184-4d88-a631-0c27381df78e.png)
![normalizedoutputbytesdistribution_onlygpu](https://user-images.githubusercontent.com/98180692/161988892-43e51e2f-723d-41cb-bf50-813f5a7fcbf0.png)

These are the normalized target distributions that I have. Do you think that is the source of not having good predictions?

I'm so sorry for the many questions but you are my only source of guidance.

**Reply by rusty1s** [reply]
Yes, you can also compute the loss of the different targets individually, in which case you can also re-weight them based on importance (especially since the follow different distributions (normal and power-law). I'm not sure why you are running OOM though. It may well be that you keep some tensors in memory and PyTorch can not properly release the computation graph. Can you verify? Feel free to also post a small code snippet.

**Reply by rtwod2** [reply]
First of all thank you sooo much for everything.

When you are saying 
> re-weight them based on importance 

I don't get that much. Do you mean, by keeping the values of the power-law distribution (outputbytes) which are closer to 0, and throwing away the outliers? 

I'm not sure about that also 
> It may well be that you keep some tensors in memory and Pytorch can not properly release the computation graph.

I'm posting the training function that I'm calling from my main.py
``` python
def train_one_epoch(epoch, model, optimizer, train_pg_graphs, batch_size, pct_close):   
    trainloader = DataLoader(train_pg_graphs, batch_size=batch_size)
     epoch_train_loss = []
    model.train()
    for step, batch_graphs in enumerate(trainloader):
        start_time_training = time.time()
        optimizer.zero_grad() 

        batch_scores = model(batch_graphs)   

        loss_train_all = model.loss(batch_scores, batch_graphs.y)
        loss_train = loss_train_all.mean(dim=0)

        loss_train.backward(torch.ones_like(loss_train)) # loss_train.backward()
        optimizer.step()

        # Update tracking
        epoch_train_loss.append(loss_train) 
    epoch_train_loss = [sum(eloss)/len(epoch_train_loss) for eloss in zip(*epoch_train_loss)] # the average of each target loss
    epoch_train_loss = torch.stack(epoch_train_loss, dim=0)  
```
and the loss function:

```python
    def loss(self, scores, targets):
        loss = nn.L1Loss(reduction='none')(scores, targets)
        return loss
```

**Reply by rusty1s** [reply]
You need to detach `loss_train` from the computation graph when adding it to `epoch_train_loss`, e.g., by converting it to a Python scalar:
```python
epoch_train_loss.append(float(loss_train))
```
By re-weighting, I mean that you can re-weight the individual losses:
```python
loss_1, loss_2 = loss_train_all.mean(dim=0)
total_loss = loss_1 + 4 * loss_2  # Weight loss_2 four times more
```

**Reply by rtwod2** [reply]
Hello, indeed detaching was the solution. I ran some experiments, e.g., by computing the loss of the different targets individually, as I wrote in the code segment above, by taking the median loss of the output bytes (power-law distributed target) and the corresponding loss of the execution time, and also by re-weighting the individual losses as you mentioned.

- By re-weighting the individual losses I had big loss, of the order of 1.6 
( I did it as you suggested 
> loss_1, loss_2 = loss_train_all.mean(dim=0)
total_loss = loss_1 + 4 * loss_2  # Weight loss_2 four times more)

- by the second one using the median loss of the power-law distributed target and the corresponding loss of the normal distribute target I have a loss 
Epoch: 0299 ['loss_train: 0.0029', 'loss_train: 0.8171']
Epoch: 0299 ['loss_eval: 0.0023', 'loss_eval: 0.9885']

- finally by computing the loss of the different targets individually and taking it the mean out of each as in code snippet above, I have a loss 

Epoch: 0299 ['loss_train: 0.0048', 'loss_train: 0.8215']
Epoch: 0299 ['loss_eval: 0.0021', 'loss_eval: 0.9812'] 

And the plots for the last one is:

![train_valid_loss_execution](https://user-images.githubusercontent.com/98180692/162187481-c6bcdfaf-1682-4113-95e4-58c27156e2cd.png)

![train_valid_loss_outputbytes](https://user-images.githubusercontent.com/98180692/162187509-4ced5532-6119-4ddf-a0e5-e89e73990e3b.png)

As you can see for the execution time (normal distributed one) it seems fine. But for the outputbytes there's no improvement at all. 
How could I deal with this?

**Reply by rtwod2** [reply]
Hi, I'm sorry for writing you back, but did you have the time to look at the above problem? Because I'm stuck at the moment on the bad predictions of the second variable. Do you have any ideas on how could I deal with it?

> Hello, indeed detaching was the solution. I ran some experiments, e.g., by computing the loss of the different targets individually, as I wrote in the code segment above, by taking the median loss of the output bytes (power-law distributed target) and the corresponding loss of the execution time, and also by re-weighting the individual losses as you mentioned.
> 
> * By re-weighting the individual losses I had big loss, of the order of 1.6
>   ( I did it as you suggested
> 
> > loss_1, loss_2 = loss_train_all.mean(dim=0)
> > total_loss = loss_1 + 4 * loss_2  # Weight loss_2 four times more)
> 
> * by the second one using the median loss of the power-law distributed target and the corresponding loss of the normal distribute target I have a loss
>   Epoch: 0299 ['loss_train: 0.0029', 'loss_train: 0.8171']
>   Epoch: 0299 ['loss_eval: 0.0023', 'loss_eval: 0.9885']
> * finally by computing the loss of the different targets individually and taking it the mean out of each as in code snippet above, I have a loss
> 
> Epoch: 0299 ['loss_train: 0.0048', 'loss_train: 0.8215'] Epoch: 0299 ['loss_eval: 0.0021', 'loss_eval: 0.9812']
> 
> And the plots for the last one is:
> 
> ![train_valid_loss_execution](https://user-images.githubusercontent.com/98180692/162187481-c6bcdfaf-1682-4113-95e4-58c27156e2cd.png)
> 
> ![train_valid_loss_outputbytes](https://user-images.githubusercontent.com/98180692/162187509-4ced5532-6119-4ddf-a0e5-e89e73990e3b.png)
> 
> As you can see for the execution time (normal distributed one) it seems fine. But for the outputbytes there's no improvement at all. How could I deal with this?

**Reply by rusty1s** [reply]
This looks like there is zero correlation between the training and validation labels of the second variable. Do you know if there might be some out-of-distribution prediction going on for validation examples?

**Reply by rtwod2** [reply]
I'm not sure about that. I found the distribution of the second variable, for the **training labels** I have

![outputtrain_distribution](https://user-images.githubusercontent.com/98180692/163975509-1649834b-c379-4b4e-ba57-fbc56ee43ba9.png)
 
for the **validation labels** I have:
![outputval_distribution](https://user-images.githubusercontent.com/98180692/163975846-317570fd-6990-4723-addc-cdad39f757f4.png)

and for the **test labels** I have:

![outputtest_distribution](https://user-images.githubusercontent.com/98180692/163976019-88f9a64b-d3d8-40a5-8ea6-e675bd7888c2.png)

How else could I check for OOD prediction?

**Reply by rusty1s** [reply]
One additional option to check for OOD is to compare the embeddings of train/val/test graphs produced by the GNN. Ideally, you would want these embeddings to be in the same space.

Nonetheless, this might only give you a better understanding of the problem, but doesn't fix your problem in the end. It is hard to give you advice without knowing your data in detail. One thing you should ask yourself is whether the model is really able to generalize and learn meaningful signals from the data to predict the correct test label (and what might be missing to achieve that). Currently, the model looks like all is does is trying to remember the training data.
